### PR TITLE
fixing issue #20818 . upgrade gnucobol upstream and the local patches to allow usage of indexed files by libdb.

### PR DESCRIPTION
**Pull request by alexbodn**
the patches to packages/gnucobol solve the problem in issue #20818  entirely, while also upgrading upstream gnucobol.

however, i'm about to ask upstream to adopt one hefty patch file, so that the package will be termux related only.

does my PR include commits out of packages/gnucobol they may be safely discarded as unrelated.

### CODE REVIEW

**Comment by alexbodn** [comment]
i know the patch seems a mess, since the commits are my progress with building and checking the package.
here is one file patch that diffs end - start without any back and forth, or other directories.
[gnucobol.patch](https://github.com/user-attachments/files/16393931/gnucobol.patch)

**Comment by GitMensch** [comment]
These are too many commits. Can you please squash them into two (one for the update, the other for bdb compatibility and use)?

File: packages/gnucobol/build.sh [hunk 1]
```diff
@@ -2,20 +2,36 @@ TERMUX_PKG_HOMEPAGE=https://gnucobol.sourceforge.io/
 TERMUX_PKG_DESCRIPTION="A free/libre COBOL compiler"
 TERMUX_PKG_LICENSE="GPL-3.0, LGPL-3.0"
 TERMUX_PKG_MAINTAINER="@termux"
-TERMUX_PKG_VERSION=3.1.2
-TERMUX_PKG_REVISION=3
+TERMUX_PKG_VERSION=3.2
+TERMUX_PKG_REVISION=4
```
Comment (GitMensch) [inline packages/gnucobol/build.sh:6]: If you increase the version then revision is to be set to 0 or 1 (please check the docs)

**Review by GitMensch — COMMENTED** [review]
Packages libdb should be part of a separate PR, no?

**Comment by alexbodn** [comment]
> Packages libdb should be part of a separate PR, no?

sure, i didn't know how to proceed, so i have initialy pushed all my commits during this period.
to correct this, i have attached to my other comment the diff of packages/gnucobol only from my first checkout to my last commit.
btw, @GitMensch , my upstream.patch doesn't have much to do with termux, has some missed missing includes and could help with any cross compilation, so i'd encourage you to apply to your distro. configure is autogenerated, but it's needed for now already.

**Comment by alexbodn** [comment]
> These are too many commits. Can you please squash them into two (one for the update, the other for bdb compatibility and use)?

sure, the workflow is triggered by commits, so all my trials had to be commited.
i've also sent a one patch from start to end.

[here it is](https://github.com/termux/termux-packages/pull/20961#issuecomment-2253042099)

> gnucobol.patch

**Comment by TomJo2000** [comment]
> sure, the workflow is triggered by commits, so all my trials had to be commited. i've also sent a one patch from start to end.

The workflows are also dispatched when force pushing to a PR branch.
I recommend you familiarize yourself with the `git commit --amend`, `git rebase -i` and `git push --force-with-lease=` commands.

**Review by TomJo2000 — CHANGES_REQUESTED** [review]
Completely unmergable in its current state.
Barely reviewable as is too.

File: scripts/lint-packages.sh [hunk 2]
```diff
@@ -259,7 +259,7 @@ lint_package() {
 					echo "PASS"
 				else
 					echo "INVALID"
-					pkg_lint_error=true
+##					pkg_lint_error=true
```
[the remark this answers is not in the document]
Comment (alexbodn) [reply scripts/lint-packages.sh:262]: this was for another package. this one does pass.
could you take a lokk a my gnucobol.patch file attached to a comment above?
it doesn't contain other directories, nor back and forth trials.

File: packages/libdb6/build.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
+TERMUX_PKG_HOMEPAGE=https://www.oracle.com/database/berkeley-db
+TERMUX_PKG_DESCRIPTION="The Berkeley DB embedded database system (library) last non AGPL"
+TERMUX_PKG_LICENSE=Sleepycat
```
[the remark this answers is not in the document]
Comment (alexbodn) [reply packages/libdb6/build.sh:3]: it's not mine. probably came when i synced my fork.

File: packages/libdb6/build.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=6.0.19
+TERMUX_PKG_REVISION=1
+TERMUX_PKG_SRCURL=https://download.oracle.com/berkeley-db/db-${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=2917c28f60903908c2ca4587ded1363b812c4e830a5326aaa77c9879d13ae18e
+TERMUX_PKG_DEPENDS="libc++"
+TERMUX_PKG_CONFLICTS="libdb-dev,libdb"
+TERMUX_PKG_REPLACES="libdb-dev,libdb"
+TERMUX_PKG_PROVIDES="libdb"
+TERMUX_PKG_BLACKLISTED_ARCHES="arm, i686, x86_64"
```
[the remark this answers is not in the document]
Comment (alexbodn) [reply packages/libdb6/build.sh:17]: sure.
first, it's another package i tried to build.
it's build took a lot of time, and i could only try on aarch64.

File: packages/libdb6/build.sh [hunk 5]
```diff
@@ -0,0 +1,38 @@
+TERMUX_PKG_HOMEPAGE=https://www.oracle.com/database/berkeley-db
+TERMUX_PKG_DESCRIPTION="The Berkeley DB embedded database system (library) last non AGPL"
```
[the remark this answers is not in the document]
Comment (alexbodn) [reply packages/libdb6/build.sh:2]: libdb6 is not the issue in this PR.
1. libdb 6.0.19 was the last libdb that is not AGPL.
2. v6 had more features dropped in very 18.1.40, like sql and client-server interfaces.
3. debian also prefer v5, even older.

File: packages/gnucobol/upstream.patch
[the remark this answers is not in the document]
Comment (alexbodn) [reply packages/gnucobol/upstream.patch:1]: i'd rather look at gnucobol.patch instead. do Ctrl-F here. it contains only changes to packages/gnucobol.

the PR itself mirrors all my commits to my termux-packages fork during my work on gnucobol. as the CI workflow would only work on commits, i had to commit all my trials until i got the problem fixed. a big thank goes to github for all these resources.

the problem my patch came to solve was with the upstream configure script that wouldn't work correctly when cross compiling.
most of my patch, the file `packages/gnucobol/upstream.patch` is directed to the upstream developers. i will delete it here once they apply it.
however it's here to make gnucobol run correctly already, based on their current official released archive.

File: packages/libdb6/build.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexbodn) [reply packages/libdb6/build.sh:3]: Sleepycat was on the list of licenses on termux-packages, but indeed lint didn't pass.

File: packages/gnucobol/upstream.patch
[the remark this answers is not in the document]
Comment (GitMensch) [reply packages/gnucobol/upstream.patch:1]: I'll have a look at this patch (and guess most things area already solved in the current dev snapshot for GnuCOBOL 3.3, but the final release will take some months)

**Comment by alexbodn** [comment]
the problems in this PR challenged me to try a better way to wrap my changes.

thank you everybody.

i'm closing this PR and have submitted another [at](https://github.com/termux/termux-packages/pull/20968).

**Comment by GitMensch** [comment]
Can you please create another PR for BDB?

**Pull request closed without merging**